### PR TITLE
Add filter command

### LINKED ISSUE #128: Implement filter feature

### PR DESCRIPTION
**Pull request by wowwwp**
Implement a filter command that allows users to filter reservations between two dates

To implement filter command,
* Add relevant syntax in the CliSyntax file
* Implement FilterCommand and FilterCommandParser
* Implement ReservationBetweenDatePredicate for filtering reservations
* Move arePrefixesPresent method to ParserUtil to reduce code
duplication
* Add isBetween method in DateTime for filtering

Fix #128

### CODE REVIEW

File: src/main/java/seedu/reserve/logic/commands/FilterCommand.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        model.updateFilteredReservationList(predicate);
+        if (model.getFilteredReservationList().isEmpty()) {
+            return new CommandResult(MESSAGE_NO_RESERVATIONS);
+        }
+        return new CommandResult(MESSAGE_SUCCESS);
```
Comment (JohannsenLum) [inline src/main/java/seedu/reserve/logic/commands/FilterCommand.java:44]: Maybe can use assertion here to asser that model.getFilteredReservationList is not empty

File: src/main/java/seedu/reserve/logic/parser/FilterCommandParser.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        DateTime startDate = ParserUtil.parseDateTimeFilter(argMultimap.getValue(PREFIX_START_DATE).get());
+        DateTime endDate = ParserUtil.parseDateTimeFilter(argMultimap.getValue(PREFIX_END_DATE).get());
+        // Checks if the start date is after the end date
+        if (startDate.compareTo(endDate) > 0) {
+            throw new ParseException(MESSAGE_INVALID_ORDER);
+        }
+
+        return new FilterCommand(startDate, endDate);
```
Comment (JohannsenLum) [inline src/main/java/seedu/reserve/logic/parser/FilterCommandParser.java:42]: Can assert here as well startDate < EndDate

File: src/main/java/seedu/reserve/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -143,4 +145,39 @@ public static DateTime parseDateTime(String dateTime) throws ParseException {
...
 
+    /**
+     * Parses a {@code dateTimeFile} to a {@code  DateTime}. Does not check if {@code dateTimeFile}
+     * is after today's date.
+     *
+     * @param dateTimeFile
+     * @return DateTime object of {@code dateTimeFile}
+     * @throws ParseException if the given {@code dateTimeFile} is invalid
+     */
+    public static DateTime parseDateTimeFilter(String dateTimeFile) throws ParseException {
```
Comment (JohannsenLum) [inline src/main/java/seedu/reserve/logic/parser/ParserUtil.java:156]: Nice JavaDocs

File: src/test/java/seedu/reserve/logic/commands/FilterCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+
+    @Test
+    public void toStringTest() {
+        ReservationBetweenDatePredicate predicate = new ReservationBetweenDatePredicate(startDate, endDate);
+        FilterCommand filterCommand = new FilterCommand(startDate, endDate);
+        String expected = FilterCommand.class.getCanonicalName() + "{predicate=" + predicate + "}";
+        assertEquals(expected, filterCommand.toString());
+    }
```
Comment (JohannsenLum) [inline src/test/java/seedu/reserve/logic/commands/FilterCommandTest.java:76]: Might want to add in more testcases to get a 100% coverage

**Review by JohannsenLum — COMMENTED** [review]
Overall Good job, just some minor things to take note of, and also remember to pull latest branch before merging

File: src/main/java/seedu/reserve/logic/commands/FilterCommand.java (hunk 1 above)
Comment (wowwwp) [reply src/main/java/seedu/reserve/logic/commands/FilterCommand.java]: resolved

File: src/main/java/seedu/reserve/logic/parser/FilterCommandParser.java (hunk 2 above)
Comment (wowwwp) [reply src/main/java/seedu/reserve/logic/parser/FilterCommandParser.java]: added

File: src/test/java/seedu/reserve/logic/commands/FilterCommandTest.java (hunk 4 above)
Comment (wowwwp) [reply src/test/java/seedu/reserve/logic/commands/FilterCommandTest.java:76]: i will do it after the other date constraint branch gets merged

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -248,6 +248,34 @@ Use case ends.
...
+Use case ends
+
+**Extensions**
+* 1a. The user enters an invalid command.
+
+  * 1a1 ReserveMate prompts user to input a valid command.
+  
+  Use case resumes at step 1.
+
+* 1b. The date entered by user is invalid.
```
Comment (basil-boh) [inline docs/DeveloperGuide.md:268]: Maybe can specify what is considered as an invalid date to avoid ambiguity!

File: docs/UserGuide.md [hunk 6]
```diff
@@ -193,6 +195,21 @@ Clears all entries from the ReserveMate
...
+* Filters all reservations between the given `DATE_TIME`, inclusive of the `DATE_TIME` provided.
+* `DATE_TIME` provided must be valid
+* The `DATE_TIME` provided for `sd/` must be before the date and time provided for `ed/`
+
+Examples:
+* `filter sd/ 2026-12-12 1400 ed/ 2026-12-15 1400` filters all reservations between 12/12/2026 1400 and 15/12/2026 1400
+
+![filter_command.png](images/filterCommand.png)
+
 ### Exiting the program : `exit`
```
Comment (basil-boh) [inline docs/UserGuide.md:213]: nicely written!

File: src/test/java/seedu/reserve/logic/parser/ReserveMateParserTest.java [hunk 7]
```diff
@@ -147,6 +149,23 @@ public void parseCommand_unrecognisedInput_throwsParseException() {
...
+        assertEquals(new FilterCommand(startTime, endTime), command);
+    }
+
+    @Test
+    public void parseCommand_filter_failure() throws Exception {
+        assertThrows(ParseException.class, () -> parser.parseCommand("filter "));
+        assertThrows(ParseException.class, () -> parser.parseCommand("filter sd/ ed/ "));
+        assertThrows(ParseException.class, () -> parser.parseCommand("filter sd/ 2025-12-13 1800 ed/"));
+        assertThrows(ParseException.class, () -> parser.parseCommand("filter ed/ 2025-12-13 1900 "));
+    }
```
Comment (basil-boh) [inline src/test/java/seedu/reserve/logic/parser/ReserveMateParserTest.java:167]: Maybe can create static string final instead of explicitly typing since the same dates are being reused to improve readability and processing!

File: src/test/java/seedu/reserve/model/reservation/DateTimeTest.java [hunk 8]
```diff
@@ -109,4 +113,18 @@ public void equals() {
         // different date and time value -> return false
         assertFalse(dateTime.equals(new DateTime(formattedDayAfterDateTime)));
     }
+
```
Comment (basil-boh) [inline src/test/java/seedu/reserve/model/reservation/DateTimeTest.java:116]: Maybe can write a test case of isSame() test to see how the command will handle boundary values

**Comment by basil-boh** [comment]
Everything else LGTM!!!

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (JohannsenLum) [reply docs/DeveloperGuide.md:268]: I think maybe no need to specify as that might be considered under attributes already which a use case should not have!

**Review by JohannsenLum — APPROVED** [review]
Do rmb to pull the master branch, merge into your branch and push the newly updated branch to resolve the conflicts
LGTM!

File: src/test/java/seedu/reserve/model/reservation/DateTimeTest.java (hunk 8 above)
Comment (wowwwp) [reply src/test/java/seedu/reserve/model/reservation/DateTimeTest.java:116]: i think there is one already it checks if its the same date
